[core-http(s)] Reuse DefaultHttpClient between ServiceClient instances

xirzec opened the pull request:
Today `ServiceClient` ends up creating a new instance of `DefaultHttpClient` each time one is not passed in via `ServiceClientOptions`. Since it is unusual for a custom `HttpClient` to be passed in, this means we recreate the `HttpClient` each time a someone creates a convenience client that wraps a generated client.

Normally, this isn't a big deal since clients are heavy enough to be cached/re-used by customer code. However, a KeyVault customer ran into a scenario where they had to recreate `CryptographyClient` many times (i.e. 5 times per second) and they noticed some pretty bad performance characteristics (out of memory, out of sockets, etc.)

While there is still more we can do with improving `CryptographyClient` to perhaps not need to be constructed so often (or making it cache its own copy of `KeyVaultClient` internally), this PR aims to solve the biggest issue with the above scenario: many copies of `DefaultHttpClient` each allocating NodeJS `http.Agent`s that then are tied to real OS resources that take a while to cleanup.

Another related note is that the Storage packages already do this today in their custom Pipeline implementation: https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/storage/storage-blob/src/utils/cache.ts

chradek reviewed (approved):
This looks great to me! Do we need a changelog entry?

(Also is it worth writing a test that makes sure that 2 service clients reference the same HttpClient if they both use the default one?)

xirzec commented:
> This looks great to me! Do we need a changelog entry?

Good point, added one so folks know about this in case it causes any weird behavior. It seems very unlikely, but doesn't hurt to document the change.

> 
> (Also is it worth writing a test that makes sure that 2 service clients reference the same HttpClient if they both use the default one?)

Added a simple test for this too!

mikeharder commented:
@HarshaNalluru was going to run perf tests before and after this change  I recommend comparing the perf of 3 apps executing requests in parallel:

1. Share a single client
2. Share a moderate number of clients (one per degree of parallelism)
3. Create a new client for each request

I would expect the steady-state throughput of (1) and (2) to be the same both before and after the change, since once things are warmed up there shouldn't be much difference between 1 or 32 clients.  The 32 clients may use a few more HTTP connections but both should perform well.

For the app which creates a new client for each operation, before this change I would expect throughput to be much lower (since each request creates a new HTTP connection), and the app may eventually fail due to client port exhaustion (if requests are created quickly enough).  After this change, I would expect app (3) to have the same perf as (1) and (2).

As long as this change doesn't cause any regressions (perf or functional), I think it's a pure win.  Helps perf in many scenarios with no downside.

mikeharder commented:
To clarify, I think it's fine to merge this PR before the perf tests are done, but I do think we should run perf tests before and after to validate this change.

HarshaNalluru commented on sdk/core/core-client/src/serviceClient.ts:
```diff
@@ -104,7 +104,7 @@ export class ServiceClient {
     this._requestContentType = options.requestContentType;
     this._baseUri = options.baseUri;
-    this._httpsClient = options.httpsClient || new DefaultHttpsClient();
+    this._httpsClient = options.httpsClient || getCachedDefaultHttpsClient();
```
I couldn't see any perceivable difference in performance when I ran the tests at https://github.com/Azure/azure-sdk-for-js/commit/6bb44f9ce5fcab9a092a5257dd294a21735d4330 with smaller durations such as 10, 30, 60 seconds.

mikeharder replied:
Did you compare the results before and after the PR?  After the PR I expected all 3 scenarios to have similar throughput, but before the PR I expect the scenario which creates a new client for each request to be much slower and it might even crash due to client port exhaustion.

HarshaNalluru replied:
[Discussed Offline, mentioning here for visibility]

These numbers are before Jeff's PR.
One more thing I realized is that the storage libraries rely on the caching already, which might have been the reason that I didn't see any difference.
Right now, updating the new tests to use core-http client directly to be able to test the fix.

HarshaNalluru replied:
Update:
There is a four fold improvement in the "new client per request" case in terms of the number of operations per second on my local machine.


Before the fix | Size (bytes) | Parallel | Duration | Warmup(Ops/sec) | Iter 1(Ops/sec) | Iter 2(Ops/sec) | Iter 3(Ops/sec)
-- | -- | -- | -- | -- | -- | -- | --
CoreHTTPSingleClientTest | 10240 | 16 | 30 | 139.71 | 163.10 | 165.10 | 163.53
CoreHTTPMultiClientsInParallelTest | 10240 | 16 | 30 | 159.72 | 161.41 | 162.01 | 161.63
CoreHTTPNewClientPerReqTest | 10240 | 16 | 30 | 39.39 | 39.74 | 39.85 | 39.67


After the fix | Size (bytes) | Parallel | Duration | Warmup(Ops/sec) | Iter 1(Ops/sec) | Iter 2(Ops/sec) | Iter 3(Ops/sec)
-- | -- | -- | -- | -- | -- | -- | --
CoreHTTPSingleClientTest | 10240 | 16 | 30 | 149.76 | 159.63 | 157.00 | 159.11
CoreHTTPMultiClientsInParallelTest | 10240 | 16 | 30 | 158.98 | 152.44 | 161.63 | 161.34
CoreHTTPNewClientPerReqTest | 10240 | 16 | 30 | 157.99 | 160.39 | 160.77 | 161.24


Before - https://github.com/HarshaNalluru/azure-sdk-for-js/tree/harshan/perf/jeff-fix-test-before
After - https://github.com/HarshaNalluru/azure-sdk-for-js/tree/harshan/perf/jeff-fix-test-after

joheredi reviewed (approved):
Looks good!
